sound/soc/sof/loader.c

```diff
@@ -357,6 +358,29 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man_parse_ipc4(struct snd_sof_dev *sdev,
+				    const struct firmware *fw)
+{
+	//TODO: parse ipc4 firmware binary
+	return 0;
+}
+
+static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+				    const struct firmware *fw)
+{
+	const struct sof_ipc4_ext_manifest_hdr *hdr;
+
+	hdr = (struct sof_ipc4_ext_manifest_hdr *)fw->data;
+	if (hdr->id == SOF_IPC4_EXT_MAN_MAGIC_NUMBER) {
+		dev_dbg(sdev->dev, "%s IPC4 fw is detected\n", __func__);
+		return snd_sof_fw_ext_man_parse_ipc4(sdev, fw);
+	}
+
+	/* fallback to ipc3 */
+	return snd_sof_fw_ext_man_parse_ipc3(sdev, fw);
```

Comment:
@plbossart @ranj063 @bardliao @ujfalusi I'm wondering should we make the extended manifest parsing completely DSP implementation specific. I.e. IPC4 would be common code, but how to extract metadata from the FW binary file (and decide whether driver should use ABI_MAJ=4 IPC language with this FW), would be behind DSP ops.

Or we just keep a set of magic variants supported in loader. E.g. already now IMX driver will call ext parse, but if SOF_EXT_MAN_MAGIC_NUMBER is not found, nothing is parsed. But if this is Intel only, probably the code should be moved away from loader.c.

Reply:
@kv2019i, with IPC3 we have platform specific extended manifest parsing for `SOF_EXT_MAN_ELEM_PLATFORM_CONFIG_DATA` type. 
Will _all_ IPC4 going to be Intel specific? Are there going to be generic information encoded in the manifest?

It does not make sense to postfix the firmware manifest parsing with ipc3 and ipc4, these are not IPC structures or are they? It is just a firmware file binary format.

I think we are mixing the IPC and firmware data structures here, which is going to cause lots of confusion and it is going to be even worst when we will have IPC5 (?) and so on, but the firmware file structure is not changing.

Reply:
> @kv2019i, with IPC3 we have platform specific extended manifest parsing for `SOF_EXT_MAN_ELEM_PLATFORM_CONFIG_DATA` type.
> Will _all_ IPC4 going to be Intel specific? Are there going to be generic information encoded in the manifest?

I don't think IPC4 is Intel specific. Anyone can use it.

> 
> It does not make sense to postfix the firmware manifest parsing with ipc3 and ipc4, these are not IPC structures or are they? It is just a firmware file binary format.

I feel the same, too. But naming is really hard to me. I have no idea what should we call to separate those two firmware file binary formats.

> 
> I think we are mixing the IPC and firmware data structures here, which is going to cause lots of confusion and it is going to be even worst when we will have IPC5 (?) and so on, but the firmware file structure is not changing.

Agree.